torch/utils/benchmark/utils/compile.py

```diff
@@ -0,0 +1,175 @@
+import torch
+
+__all__ = ["bench_all", "benchmark_compile"]
+
+import torch._dynamo
+from torch.utils.benchmark import Timer
+
+from typing import Optional, List, Callable, Union, Any, cast
+
+try:
+    from tabulate import tabulate
+    HAS_TABULATE = True
+except ImportError:
+    HAS_TABULATE = False
+
+_warned_tensor_cores = False
+_default_float_32_precision = torch.get_float32_matmul_precision()
+
+
+if HAS_TABULATE:
+    def _enable_tensor_cores():
+        global _warned_tensor_cores
+
+        if torch.cuda.is_available():
+            if torch.backends.cuda.matmul.allow_tf32 is False and torch.cuda.get_device_capability() >= (8, 0):
+                torch.set_float32_matmul_precision("high")
+                if not _warned_tensor_cores:
+                    print("Your GPU supports tensor cores")
+                    print("we will enable it automatically by setting `torch.set_float32_matmul_precision('high')`")
+                    _warned_tensor_cores = True
+
+    def _disable_tensor_cores():
+        torch.set_float32_matmul_precision(_default_float_32_precision)
+
+    def bench_loop(
+        model: Union[torch.nn.Module, Callable],
+        sample_input: Union[torch.Tensor, Any],
+        num_iters: int = 5,
+        optimizer: torch.optim.Optimizer = None,
+        loss_fn: Callable = None,
+    ):
+        # Define the statement and setup for the benchmark
+        if optimizer and loss_fn:
+            # Training mode
+            stmt = """
+    output = model(sample_input)
+    loss = loss_fn(output) if loss_fn else output.sum()
+    loss.backward()
+    optimizer.step()
+    optimizer.zero_grad()
+            """
+        else:
+            # Inference mode
+            stmt = "model(sample_input)"
+
+        # Create the Timer object
+        timer = Timer(
```

Comment:
Ed gave us some interesting feedback on this, the compilation time is eaten every time so we have a few workarounds
1. Add a warmup to timer object so the compilation is ignored but only for when we're not explicitly measuring compilations
2. quiescence mode via torch._dynamo.run can help us pass the dynamo cache as a global variable instead
3. In tests we can use CompileCounter to explicitly make sure that no new compilation is happening

Reply:
So for this tool, I think we should (1) measure how long it takes to compile and report that, but (2) error if we end up recompiling on subsequent iteration loops